### DISCUSSION TITLE
Peer reconnects

### DISCUSSION THREAD

**Discussion by elgatito** [arvidn/libtorrent#6344] [Q&A]
Hello,

When torrent is started, libtorrent is opening connections to peers, then I set pieces to download, but if user is stopping consuming data, we have a situation when we have all the requested pieces for some time.
And then I see that we have less and less peer connections. 
As I understand, remote party is closing connections due to idle time or any other reason (`close_redundant_connections` is disabled).

The question is how to reconnect to peers? Is there such possibility?

For non-crowded torrents it can become a problem when we have few peers and later we have only one, and the only way out is to drop/add torrent again to reconnect to everyone.

**Comment by arvidn** [comment]
I don't think there's a specific trigger when the torrent transitions from "finished" -> "downloading" mode, the way there is when transitioning from "paused" -> "non-paused", to try to reconnect to peers. However, torrents with fewer peers than the target is always trying to find peers to connect to. However, there's a cool-off period for peers we've just been connected to.

Do you have a low `connection_speed`? That setting controls the rate at which we attempt new connections.

Normally, when resuming a torrent, there's also a connection-speed boost, where we allocate extra outgoing connections slots to the torrent. Perhaps there should be such a boost for transitioning from "finished" -> "downloading" as well.

**Reply by elgatito** [reply]
Thanks for replying, @arvidn .

I have big connection speed, initial connections are fine as well as the ongoing if data reader is constantly getting data, so peer connections are not idle.

I'm more writing about situations when we have at most 5 peers, and when we have nothing to download for long time, statistics show 0/1/2 peers connected (number of available peers is still the same), and if I mark another bunch of pieces for downloading, status is changing to `downloading`, but it looks like we do not reconnect to peers that we were using previously.

That was my question, do we reconnect to peers, that were fine previously, but then they disconnected from us?

**Reply by arvidn** [reply]
yes, but there's a "cool-off" period. If we were connected to them very recently, we won't reconnect for some time. I think it may be in the order of minutes. Also, even if you aren't disconnecting redundant peers, the other peers might.

**Reply by AllSeeingEyeTolledEweSew** [reply]
> Also, even if you aren't disconnecting redundant peers, the other peers might.

Is there a way to avoid this? Is this a consequence of us sending a "not interested" message, and could we avoid sending it?

**Reply by elgatito** [reply]
> I think it may be in the order of minutes.

@arvidn Can you point to the code where re-connect logic is? Because when I look on number of connected peers, they are decreasing if there is nothing to download, but never increasing. Again, when I have at most 3-5 peers for specific torrent, so there are no other options.
